Consistently use @raise tags in Stdlib docs

Et7f3 opened the pull request:
All comment have the same shape
`(**
   comment
 *)`
Use of @raise tags when it was in text.
Do you think it worth ask for ocamldoc autodetect exception and try to detect which value cause error (by scanning code like if match ... with structure ?)

Et7f3 committed:
fix formatting comment for ListLabels.mli

All comment have the same shape
`(**
   comment
 *)`
Use of @raise tags when it was in text.
Do you think it worth ask for ocamldoc autodetect exception and try to detect which value cause error (by scanning code like if match ... with structure ?)

dra27 committed:
Corrections

dra27 commented:
I like the conversion to `@raise`, thanks (would you be happy to do similar in `stdlib/list.mli`?

I'm less thrilled about the necessity of reformatting double-spaces and exactly where `*)` marks appear.

Et7f3 committed:
Use of @raise tags for list.mli

Use of @raise tags when it was in text. for list.mli file.
Like asked by @dra27 in https://github.com/ocaml/ocaml/pull/8644

Et7f3 commented:
I have not modified double space in list.mli and change where `*)` mark appear.

Et7f3 commented:
some function like `combine` don't have the value of Invalid_argument.
I may need to fix it 🤔 .
![image](https://user-images.githubusercontent.com/29592775/56772882-1aa45280-67bc-11e9-91c3-6b9014a3186d.png)
I will use grep for finding other examples in all the code.

gasche commented:
> I may need to fix it.

Please don't; the string payload of Invalid_argument occasionally changes, and it's a bad idea for users to rely on it.

dra27 committed:
More corrections

dra27 commented:
Please do try to test these - `make world` builds documentation by default and you can always do `make -C ocamldoc html_doc` to test changes.

dra27 commented:
I leave it to @gasche to rule whether the whitespace alterations in `listLabels.mli` should be reverted before this is squashed, but it's otherwise good to go.

It's entirely up to you, @Et7f3, but a quick grep reveals arrayLabels.mli, array.mli, bigarray.mli, buffer.mli, bytesLabels.mli, bytes.mli, camlinternalFormat.ml, char.mli, digest.mli, filename.mli, float.mli, gc.mli, int32.mli, int64.mli, nativeint.mli, queue.mli, scanf.mli, stack.mli, stdlib.mli, stringLabels.mli, string.mli, sys.mli and weak.mli also contain "Raise", if you fancied changing some more.

A further improvement which could be applied here and more widely would be to remove the arguments in documentation comments from `Invalid_argument` and `Failure`. Arguably that should have been part of the change to the `[@ocaml.warn_on_literal_pattern]` attribute

Et7f3 commented:
The modules [CamlinternalFormat](https://caml.inria.fr/pub/docs/manual-ocaml/libref/CamlinternalFormat.html) have a manual page and comment but comment is not displayed. Is it intended ? Or we should transform comment to ocamldoc format ?

How should I treat `Raise {!Scanf.Scan_failure}` @raise don't support it.

Et7f3 commented:
how should I treat `Raise {!Scanf.Scan_failure}` @raise don't support it.

Et7f3 committed:
Use of @raise tags for int64.mli

Et7f3 committed:
Use of @raise tags for int32.mli

Et7f3 committed:
Use of @raise tags for nativeint.mli

gasche commented:
Unfortunately, I don't have the time right now to review this properly. @dra27, if you do have the time at some point, I would be happy to defer all decisions to you.

damiendoligez commented:
@dra27 the whitespace changes shouldn't prevent us from merging this.

@Et7f3:
* CamlinternalFormat has a comment at the top that says `(* No comments, OCaml stdlib internal use only. *)`. Please do not convert the comments.
* `Raise {!Scanf.Scan_failure}`: are there many cases of this? If not, I would just drop the prefix. If there are many, you should keep them as-is until we add the feature to ocamldoc.
* You'll need to rebase this PR as it has conflicts with the current trunk.

dra27 commented on stdlib/list.mli:
```diff
@@ -52,24 +52,24 @@ val cons : 'a -> 'a list -> 'a list
-(** Return the first element of the given list. Raise
-   [Failure "hd"] if the list is empty. *)
+(** Return the first element of the given list.
+   @raise Failure ["hd"] if the list is empty. *)
```
```suggestion
   @raise Failure if the list is empty. *)
```

dra27 commented on stdlib/list.mli:
```diff
@@ -52,24 +52,24 @@ val cons : 'a -> 'a list -> 'a list
-(** Return the given list without its first element. Raise
-    [Failure "tl"] if the list is empty. *)
+(** Return the given list without its first element.
+    @raise Failure ["tl"] if the list is empty. *)
```
```suggestion
    @raise Failure if the list is empty. *)
```

dra27 commented on stdlib/list.mli:
```diff
@@ -52,24 +52,24 @@ val cons : 'a -> 'a list -> 'a list
    The first element (head of the list) is at position 0.
-   Raise [Failure "nth"] if the list is too short.
-   Raise [Invalid_argument "List.nth"] if [n] is negative. *)
+   @raise Failure ["nth"] if the list is too short.
+   @raise Invalid_argument ["List.nth"] if [n] is negative. *)
```
```suggestion
   @raise Failure if the list is too short.
   @raise Invalid_argument if [n] is negative. *)
```

dra27 commented on stdlib/list.mli:
```diff
@@ -52,24 +52,24 @@ val cons : 'a -> 'a list -> 'a list
     The first element (head of the list) is at position 0.
     Return [None] if the list is too short.
-    Raise [Invalid_argument "List.nth"] if [n] is negative.
+    @raise Invalid_argument ["List.nth"] if [n] is negative.
```
```suggestion
    @raise Invalid_argument if [n] is negative.
```

dra27 reviewed (commented):
There are four instances of `Failure` and `Invalid_argument` still with an argument (I don't think ocamldoc renders `@raise Exception ["argument"]` very nicely, though I haven't actually checked).

I'm happy for this to go in without a Changes entry, unless you want to push one (potentially with those four adjustments) and then I'll (finally) merge this, thanks!

Et7f3 committed:
Apply suggestions from code review

Co-authored-by: David Allsopp <david.allsopp@metastack.com>

Et7f3 committed:
use sed.js script published in GPR

Et7f3 committed:
fix: automatic modification

Et7f3 commented:
My sed.js used:
```
const fs = require("fs")

const my_args = process.argv.slice(2)
my_args.forEach(name => {
  const code = fs.readFileSync(name)
  fs.writeFileSync(name, code.toString().replace(/\s*Raise \[(Not_found|Invalid_argument|Division_by_zero|End_of_file|Sys_error|Failure)\]/gm, "\n   @raise $1"))
})
```

I launched with: `node sed.js $(find stdlib -name "*.mli")`

(I haven't used ocaml stdlib because I think the nodejs has more utility function and I feel dumb to have beginned this task by hand :clown_face:)
EDIT:
I have used 3 spaces because it was the number of space used in the fist file edited.

dra27 commented on stdlib/listLabels.mli:
```diff
@@ -57,60 +61,66 @@ val compare_length_with : 'a list -> len:int -> int
     The first element (head of the list) is at position 0.
     Return [None] if the list is too short.
-    Raise [Invalid_argument "List.nth"] if [n] is negative.
+    @raise Invalid_argument ["List.nth"] if [n] is negative.
```
```suggestion
    @raise Invalid_argument if [n] is negative.
```

dra27 commented on stdlib/scanf.mli:
```diff
@@ -313,7 +313,8 @@ val bscanf : Scanning.in_channel -> ('a, 'b, 'c, 'd) scanner
     - [c]: reads a single character. To test the current input character
       without reading it, specify a null field width, i.e. use
-      specification [%0c]. Raise [Invalid_argument], if the field width
+      specification [%0c].
+    - Raise [Invalid_argument], if the field width
```
This change is incorrect.

Et7f3 replied:
```suggestion
      specification [%0c]. Raise [Invalid_argument], if the field width
```
Ok I apply that.

dra27 reviewed (commented):
Nice! Two small nits remain. The use of 3 spaces before `@raise` means that the source-formatting is off in many places. It won't affect the HTML, but if you're doing it all with a script why not:
- Preprocess by hand the small number of places where `Raise` does not begin on a newline
- Capture the `\s` in the regexp and use the capture group for the spacing instead of a hard-coded 3 spaces?

Et7f3 committed:
fix(doc): Last raise by hand.

Et7f3 committed:
Apply suggestions from code review

Co-authored-by: David Allsopp <david.allsopp@metastack.com>

Et7f3 commented:
Does I modify failwith description or it is obvious ?

Et7f3 commented on stdlib/stream.mli:
```diff
@@ -67,7 +67,8 @@ val iter : ('a -> unit) -> 'a t -> unit
 (** Return the first element of the stream and remove it from the
-   stream. Raise {!Stream.Failure} if the stream is empty. *)
+   stream.
+   @raise Stream.Failure if the stream is empty. *)
```
Failure ? or Stream.Failure. Their is another one.

dra27 replied:
I would put `Stream.Failure` so that the text which will be displayed now is clear. https://github.com/ocaml/odoc/issues/440 should deal with this in odoc ultimately.

Et7f3 committed:
fix(doc): Align @raise with the paragraph.

Et7f3 commented:
> The use of 3 spaces before @raise means that the source-formatting is off in many places.

In many files 2/3 or 3/4 spaces are used.

> Capture the \s in the regexp and use the capture group for the spacing instead of a hard-coded 3 spaces?

sometime we have `text Raise` so `\s` == " "

> Preprocess by hand the small number of places where Raise does not begin on a newline

I have fixed by hand.

dra27 commented:
What I meant with the pre-processing is that if you ensure by hand that "Raise" is always on a newline then you won't have any cases of "_text_ Raise"

dra27 commented:
> Does I modify failwith description or it is obvious ?

I don't understand - where?

dra27 commented:
Oh, do you mean the description of `Stdlib.failwith` (and presumably `Stdlib.invalid_arg`). I would leave those as they are, yes, given that those functions are precisely about raising an exception.

dra27 commented on stdlib/listLabels.mli:
```diff
@@ -57,60 +61,66 @@ val compare_length_with : 'a list -> len:int -> int
    The first element (head of the list) is at position 0.
-   Raise [Failure "nth"] if the list is too short.
-   Raise [Invalid_argument "List.nth"] if [n] is negative. *)
+   @raise Failure ["nth"] if the list is too short.
+   @raise Invalid_argument ["List.nth"] if [n] is negative.
```
```suggestion
   @raise Failure if the list is too short.
   @raise Invalid_argument if [n] is negative.
```

dra27 commented on stdlib/bytes.mli:
```diff
@@ -124,33 +116,29 @@ val blit : bytes -> int -> bytes -> int -> int -> unit
-     as new byte sequence.
-
-    Raise [Invalid_argument] if the result is longer than
+    as new byte sequence.
```
```suggestion
    as a new byte sequence.
```
(it's not your typo, but while we're here!)

dra27 commented on stdlib/stdlib.mli:
```diff
@@ -361,7 +361,7 @@ external ( * ) : int -> int -> int = "%mulint"
 external ( / ) : int -> int -> int = "%divint"
 (** Integer division.
-   Raise [Division_by_zero] if the second argument is 0.
+   @raise Division_by_zero if the second argument is 0.
```
This wants to go below the paragraph (i.e. at the end of the documentation comment)

dra27 commented on stdlib/stdlib.mli:
```diff
@@ -376,7 +376,7 @@ external ( mod ) : int -> int -> int = "%modint"
    If [y = 0], [x mod y] raises [Division_by_zero].
    Note that [x mod y] is negative only if [x < 0].
-   Raise [Division_by_zero] if [y] is zero.
+   @raise Division_by_zero if [y] is zero.
```
Likewise, this wants to go after the next line

dra27 commented on stdlib/stream.mli:
```diff
@@ -67,10 +67,12 @@ val iter : ('a -> unit) -> 'a t -> unit
 
 val empty : 'a t -> unit
-(** Return [()] if the stream is empty, else raise {!Stream.Failure}. *)
+(** Return [()]
```
This function is also about expressly raising an exception - I think the description is better as it was without the `@raise` tag

Et7f3 replied:
I will force push HEAD~

dra27 commented on stdlib/stringLabels.mli:
```diff
@@ -167,17 +157,16 @@ val index_from : string -> int -> char -> int
 
     [String.index_opt s c] is equivalent to [String.index_from_opt s 0 c].
-    Raise [Invalid_argument] if [i] is not a valid position in [s].
+   @raise Invalid_argument if [i] is not a valid position in [s].
```
```suggestion
    @raise Invalid_argument if [i] is not a valid position in [s].
```
(only since all the others are so beautifully perfect!)

dra27 reviewed (commented):
This is a really nice set of changes! I went through everything one last time - there are the last few smallest nits (and, as there were some, there is a typo which I spotted but which was nothing to do with the change)

This feels highly worth of a Changes entry, if you want to add an entry in the `Changes` file in the `Standard Library` section? I'm happy to rebase this on to 4.11, so it can go in the `Standard Library` section of the 4.11 changes.

Et7f3 commented:
`grep "Raise" $(find . -name "*.mli")` show me others in otherlibs and in many places.

> I'm happy to rebase this on to 4.11

I have already rebased onto trunk.

Et7f3 committed:
Apply suggestions from code review

Co-authored-by: David Allsopp <david.allsopp@metastack.com>

dra27 commented:
> `grep "Raise" $(find . -name "*.mli")` show me others in otherlibs and in many places.

Please feel very welcome to do a follow-up PR, although this one is now quite large and I'd prefer not to have to read them all again!

> I'm happy to rebase this on to 4.11
> 
> I have already rebased onto trunk.

That's correct - this will be merged to trunk, I was talking cherry-pick it all for OCaml 4.11 (i.e. don't worry!)

Et7f3 commented:
Only read the two last commit (I just moved two lines in the last commit)

Et7f3 committed:
Add my name in Changes files.

dra27 commented on stdlib/bigarray.mli:
```diff
@@ -347,7 +347,7 @@ module Genarray :
      If [a] has Fortran layout, the coordinates must be greater or equal
      than 1 and less or equal than the corresponding dimensions of [a].
-     Raise [Invalid_argument] if the array [a] does not have exactly [N]
+     @raise Invalid_argument if the array [a] does not have exactly [N]
```
Ugh, sorry - I just ran the PR through a better diff tool and spotted that this needs to go at the end of documentation comment too!

Et7f3 committed:
oops 4.11 :sweat_smile:

dra27 commented on stdlib/float.mli:
```diff
@@ -453,9 +446,8 @@ module Array : sig
-     Raise [Invalid_argument] if [ofs] and [len] do not
+      storing [x] in elements number [ofs] to [ofs + len - 1].
+      @raise Invalid_argument if [ofs] and [len] do not
      designate a valid subarray of [a]. *)
```
```suggestion
      designate a valid subarray of [a]. *)
```

dra27 commented on stdlib/float.mli:
```diff
@@ -398,52 +398,45 @@ module Array : sig
+      containing the elements number [start] to [start + len - 1]
+      of floatarray [a].
+      @raise Invalid_argument if [start] and [len] do not
      designate a valid subarray of [a]; that is, if
      [start < 0], or [len < 0], or [start + len > length a]. *)
```
```suggestion
      designate a valid subarray of [a]; that is, if
      [start < 0], or [len < 0], or [start + len > length a]. *)
```

dra27 commented on stdlib/int32.mli:
```diff
@@ -56,7 +56,8 @@ external mul : int32 -> int32 -> int32 = "%int32_mul"
 
 external div : int32 -> int32 -> int32 = "%int32_div"
-(** Integer division.  Raise [Division_by_zero] if the second
+(** Integer division.
+   @raise Division_by_zero if the second
    argument is zero.  This division rounds the real quotient of
    its arguments towards zero, as specified for {!Stdlib.(/)}. *)
```
```suggestion
(** Integer division. This division rounds the real quotient of
   its arguments towards zero, as specified for {!Stdlib.(/)}.
   @raise Division_by_zero if the second
   argument is zero.  *)
```

dra27 commented on stdlib/lazy.mli:
```diff
@@ -63,15 +63,15 @@ external force : 'a t -> 'a = "%lazy_force"
     result.  If [x] has already been forced, [force_val x]
     returns the same value again without recomputing it.
-    Raise {!Undefined} if the forcing of [x] tries to force [x] itself
+    @raise Undefined if the forcing of [x] tries to force [x] itself
     recursively.
     If the computation of [x] raises an exception, it is unspecified
     whether [force_val x] raises the same exception or {!Undefined}.
```
```suggestion

    If the computation of [x] raises an exception, it is unspecified
    whether [force_val x] raises the same exception or {!Undefined}.
    @raise Undefined if the forcing of [x] tries to force [x] itself
    recursively.
```

Et7f3 replied:
can you remove your space before `@raise` so I can commiut from GitHub UI (and you get credited)

dra27 commented on stdlib/listLabels.mli:
```diff
@@ -36,14 +37,17 @@ type 'a t = 'a list = [] | (::) of 'a * 'a list (**)
-(** Return the first element of the given list. Raise
-   [Failure "hd"] if the list is empty. *)
+(** Return the first element of the given list.
+   @raise Failure ["hd"] if the list is empty.
```
```suggestion
   @raise Failure if the list is empty.
```

dra27 commented on stdlib/listLabels.mli:
```diff
@@ -57,60 +61,66 @@ val compare_length_with : 'a list -> len:int -> int
-(** Return the given list without its first element. Raise
-   [Failure "tl"] if the list is empty. *)
+(** Return the given list without its first element.
+   @raise Failure ["tl"] if the list is empty.
```
```suggestion
   @raise Failure if the list is empty.
```

dra27 commented on stdlib/nativeint.mli:
```diff
@@ -59,7 +59,9 @@ external mul : nativeint -> nativeint -> nativeint = "%nativeint_mul"
 
 external div : nativeint -> nativeint -> nativeint = "%nativeint_div"
-(** Integer division.  Raise [Division_by_zero] if the second
+(** Integer division.
+
+   @raise Division_by_zero if the second
    argument is zero.  This division rounds the real quotient of
    its arguments towards zero, as specified for {!Stdlib.(/)}. *)
```
```suggestion
(** Integer division. This division rounds the real quotient of
   its arguments towards zero, as specified for {!Stdlib.(/)}.

   @raise Division_by_zero if the second
   argument is zero. *)
```

dra27 commented on stdlib/scanf.mli:
```diff
@@ -448,7 +448,7 @@ val bscanf : Scanning.in_channel -> ('a, 'b, 'c, 'd) scanner
     according to the format string:
 
-    - Raise {!Scanf.Scan_failure} if the input does not match the format.
+    - Raise [Scan_failure] if the input does not match the format.
```
This reference is fine that can go back to `{!Scanf.Scan_failure}`

Et7f3 replied:
```suggestion
    - Raise {!Scanf.Scan_failure} if the input does not match the format.
```

dra27 reviewed (commented):
Sorry - I ran the diff through the `patdiff` tool which has better context display than GitHub. I promise this is the last tranche!

Et7f3 committed:
Apply suggestions from code review

Co-authored-by: David Allsopp <david.allsopp@metastack.com>

Et7f3 committed:
should be the last one but it won't :joy:

dra27 reviewed (approved):
I've checked through the HTML output as well - this is a very nice set of updates to the standard library documentation, thank you!
